internal/zero/controller/usage_reporter.go

```diff
@@ -0,0 +1,278 @@
+	userEmail       string
+	lastSignedInAt  time.Time
+}
+
+type usageReporter struct {
+	api *sdk.API
+
+	mu       sync.Mutex
+	byUserID map[string]usageReporterRecord
+	updates  map[string]struct{}
+}
+
+func newUsageReporter(api *sdk.API) *usageReporter {
+	return &usageReporter{
+		api:      api,
+		byUserID: make(map[string]usageReporterRecord),
+		updates:  make(map[string]struct{}),
+	}
+}
+
+func (ur *usageReporter) report(ctx context.Context, records []usageReporterRecord) error {
+	req := cluster.ReportUsageRequest{}
+	for _, record := range records {
+		req.Users = append(req.Users, cluster.ReportUsageUser{
+			DisplayName:    record.userDisplayName,
+			Email:          record.userEmail,
+			Id:             record.userID,
+			LastSignedInAt: record.lastSignedInAt,
+		})
+	}
+	return ur.api.ReportUsage(ctx, req)
+}
+
+func (ur *usageReporter) run(ctx context.Context, client databroker.DataBrokerServiceClient) error {
+	// first initialize the user collection
+	serverVersion, latestRecordVersion, err := ur.runInit(ctx, client)
+	if err != nil {
+		return err
+	}
+
+	// run the continuous sync calls and periodically report usage
+	return ur.runSync(ctx, client, serverVersion, latestRecordVersion)
+}
+
+func (ur *usageReporter) runInit(ctx context.Context, client databroker.DataBrokerServiceClient) (serverVersion, latestRecordVersion uint64, err error) {
+	_, _, err = syncLatestRecords(ctx, client, ur.onUpdateSession)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	serverVersion, latestRecordVersion, err = syncLatestRecords(ctx, client, ur.onUpdateUser)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return serverVersion, latestRecordVersion, nil
+}
+
+func (ur *usageReporter) runSync(ctx context.Context, client databroker.DataBrokerServiceClient, serverVersion, latestRecordVersion uint64) error {
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return syncRecords(ctx, client, serverVersion, latestRecordVersion, ur.onUpdateSession)
+	})
+	eg.Go(func() error {
+		return syncRecords(ctx, client, serverVersion, latestRecordVersion, ur.onUpdateUser)
+	})
+	eg.Go(func() error {
+		return ur.runReporter(ctx)
+	})
+	return eg.Wait()
+}
+
+func (ur *usageReporter) runReporter(ctx context.Context) error {
+	// every minute collect any updates and submit them to the API
+	timer := time.NewTicker(time.Minute)
+	defer timer.Stop()
+
+	for {
+		// collect the updated records since last run
+		ur.mu.Lock()
+		records := make([]usageReporterRecord, 0, len(ur.updates))
+		for userID := range ur.updates {
+			records = append(records, ur.byUserID[userID])
+		}
+		clear(ur.updates)
+		ur.mu.Unlock()
+
+		// report the records with a backoff in case the API is temporarily unavailable
+		if len(records) > 0 {
+			log.Info(ctx).Int("updated-users", len(records)).Msg("reporting usage")
+			err := backoff.Retry(func() error {
+				err := ur.report(ctx, records)
+				if err != nil {
+					log.Error(ctx).Err(err).Msg("error reporting usage")
```

Comment:
```suggestion
					log.Warn(ctx).Err(err).Msg("error reporting usage")
```

as it does not affect pomerium itself